Compound index support for StorageManager

poltak opened the pull request:
@ShishKabab would be good to get your review + opinions on these changes before going ahead and bringing them in.

Notable changes (more info in commit messages):

1. ~~removed `pk` bool field from collection `fields` definitions~~ (moved to `IndexDefinition`)
    - currently only ever used for ordering the indexable fields in Dexie schemas, however indexable fields in StorageManager are already declared in an ordered data structure
    - only relates to indexable fields, but was declared as part of general data `fields` schema; need to handle case of user declaring field as PK but never declaring it in `indices`
    - instead have taken inspiration from Dexie schemas by saying the first `indices` element is the primary key index (JSDoc added to typings to guide users when writing schemas)
    - updated direct-linking schema to reflect this

2. whenever accessing a `CollectionField` definition - usually via something like `collection.fields[indexName]` - we now check if the `indexName` is a compound index or not, and act slightly differently if so
    - ~~again taking inspiration from Dexie's compound index shape: expressed as "[fieldA+fieldB]". See new test case for example~~
    - new shape: compound indexes are now expressed as type `[string, string]` where each element corresponds to the name of a field - now no need to mess around with string processing. It also means the `indices` array in StorageManager collection defs is now of type `Array<string|[string, string]>` (thanks to suggestion from @ShishKabab )

TODOs (either in here or later):

- [ ] move current Dexie schema to StorageManager schema
- [ ] detection and proper errors thrown when user declares invalid schemas (I don't think these can be done at compile time): 
    - [ ] fields that appear in compound indexes should not be type `text`
    - [ ] indexes should not be made for fields of type `text`
    - [ ] `indices` array needs to contain __exactly one__ `IndexDefinition` with `pk` set

poltak commented:
@ShishKabab Earlier I updated the compound index implementation to use the nested 2-element array shape for valid compound index expressions. 
eg: `['fieldA', 'fieldB']` rather than `'[fieldA+fieldB]'`. 
Was able to avoid the string processing, and slightly simplify the surrounding code. OP updated to explain a bit more

If we don't want to go with using the order of the `indices` array to express which index is the PK, we still need to think about a way to express primary compound indexes and also express the order of fields within a compound index in our collection defs.

ShishKabab commented:
I understand better now why you got rid of the `pk` field, and it makes sense. The only thing I don't like right now is that the pk is implicit and the user might not even know it exists, or that the ordering of the fields is important. Could we maybe change the type of IndexType to the following:

```
export type IndexSourceFields = string | Array<string>
export type IndexType = IndexSourceFields | IndexDefinition
export interface IndexDefinition {
  source : IndexSourceFields
  fullTextIndexName? : string // Don't generate name automatically, but use this one
  pk? : boolean
}
```

And then just checking if the first field is a `pk`, throwing an error if not? Sorry for the nitpicking, but in my experience everything too magical piles up and makes a library difficult to understand and debug.

About the schema splitting among multiple Version.stores() calls, we're merging them all based on the version number, right?

Do I understand correctly what you mean by the naming of the index fields, with the above config option addressing that? Anyway, if this cleanup is costing you too much energy, it can definitely wait. It does have to be done at some point though, when the library will be used in other contexts, or when we want to update those core schemas, but that could well take a while.

poltak commented:
Love the proposed `IndexDefinition` shape! 

> And then just checking if the first field is a pk, throwing an error if not?

Though I don't see any need for checking this anymore, as order should no longer need to be taken into account.

One more thing I just thought of:
We might also want to add opt. flags to `IndexDefinition` for `unique` and maybe `autoInc` (depends on `pk` being set). @subrat-sahu has needed these in his indexes for the lists collection. At the moment he's just just using the Dexie syntax, which apparently works fine although I don't think it was intended in the implementation of the `dexie-schema` module. See:
https://github.com/subrat-sahu/Memex/blob/961222aa10d9920b72e2c17daf3516b173a78a29/src/custom-lists/background/storage.ts#L22

I'm writing this rushed but will have a bit of a think about how these could work in a general way, but it may be that straightforward.

> Do I understand correctly what you mean by the naming of the index fields, with the above config option addressing that?

I didn't get around to playing with the schema porting stuff today, but yes this `fullTextIndexName` option `IndexDefinition` sounds like exactly what I need with the other problem I described yesterday. I'll see where I get with that tomorrow and give you more of an update on the other schema porting stuff then.

poltak commented:
@ShishKabab I have implemented that shape + the additional unique, inc fields I talked about yesterday in 0229934.

Some minor opinionated changes I have made, which are open to discussion:
- decided to call `source` field just `field` as, while the JSDoc should explain the purpose fully, I feel it makes the collection schemas more understandable at a glance (look at direct linking + test case schema updates and see what you think)
- `indicies` is simply type `IndexDefinition[]`, rather than `IndexType[]`. Main reason being it adds a lot of messy conditional code to support three different shapes for any one element of `indices` (`string`, `string[]` and `IndexDefinition`). Effectively, it just means a user has to write each element of their `indices` array as an object with at least the `field` key defined, rather than the shortcut of just writing a string element.

One more thing that came to mind as I was implementing these different index options:
Currently we determine whether an index affords full-text search if the `CollectionField`'s `type` is set to `'text'`, while all other index options (uniqueness, PK, auto-inc PK) are handled in the `IndexDefinition`. It does maybe feel like it has more to do with the field's "type" than anything else though. 
e.g., if we moved it to an option on `IndexDefinition`, it would only work if both that option is set + the corresponding field is `'text'` type - from user POV that feels a bit unnecessarily complicated to get right.

This part of the code may give a better idea of what I mean: 
https://github.com/WorldBrain/Memex/blob/0229934d804b8952fff7a3b744e162334e0b10c2/src/search/search-index-new/storage/dexie-schema.ts#L55-L65

ShishKabab commented:
Yup, the `field` key makes sense, creates a very clear connection.

`indices` as `IndexDefinition[]` also makes sense. Maybe be slightly more to write, but in the end I also like the visual uniformity better  :)

And yeah, about the full-text search, for me using the `text` field means `text` instead of `varchar` in RDBMS terms as you mentioned. So it kinda makes sense for me to automatically make those full-text searchable if an index is created on it. We could make it more explicit, which also has it's benefits though. Though call...
